dvc/project.py

```diff
@@ -195,7 +199,38 @@ def _check_cyclic_graph(graph):
         if cycles:
             raise CyclicGraphError(cycles[0])
 
-    def add(self, fname, recursive=False):
+    @staticmethod
+    def _check_can_commit(stage, entry, force):
+        if entry.checksum and entry.changed_checksum():
+            msg = (
+                "file/directory '{}'({}) changed. Are you sure you want to "
+                "commit it?"
+            ).format(entry.rel_path, stage.relpath)
+            if not force and not prompt.confirm(msg):
+                raise CommitError(
+                    "unable to commit changed file/directory '{}'({}). Use "
+                    "`-f|--force` to force.".format(
+                        entry.rel_path, stage.relpath
+                    )
+                )
+
+    def commit(self, target, with_deps=False, recursive=False, force=False):
+        if target and not recursive:
+            stages = self._collect(target, with_deps=with_deps)
+        else:
+            stages = self.active_stages(target)
+
+        with self.state:
+            for stage in stages:
+                for entry in stage.deps + stage.outs:
+                    self._check_can_commit(stage, entry, force)
+
+            for stage in stages:
+                stage.save()
```

Comment:
can `dvc commit` change stage files?

Reply:
Yes. It will re-calculate checksums if they are missing or they don't match(this one only if explicitly confirmed with a prompt or --force flag).

Reply:
It makes sense. How about this case then. In the next version we are going o have no-commit enabled by default, right? So, let's imagine I run something in `master` and then checkout another branch. It means that checksums in stage files are wrong now. But is does not mean that I want force them when I run `dvc commit`, all I want to do is to save files to cache. It might be that we'll need to abort the checkout itself, similar to what git is doing, right? Any thoughts on this? Basically, I'm thinking is it always the case that we want to force the change in stage files or do we just want to push data (e.g. generated in a different branch, or previous git commit, or whatnot) to cache.

Reply:
That's a great point! We need to have that check on `dvc checkout`. Looking into it...

Reply:
@shcheklein Just checked and we do check that on `dvc checkout` anyway. E.g. not committed stage.dvc means that there is no cache for it and so `dvc checkout` will prompt user asking `file 'foo' is going to be removed. Are you sure you want to proceed?` So we are covered there. Still, the message is not very informative, so I'll add a clarification for it.

Reply:
For the record:created https://github.com/iterative/dvc/issues/1622 for it for now.